### PR TITLE
[BEAM-259] Enable RunnableOnService for batch.

### PR DESCRIPTION
**Pull request by amitsela**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Handle empty Flatten for bounded.

Spark will bubble out a SparkException for user code failure, so this won't catch. Asserting on the
error message should be good enough.

outputWithTimestamp should handle start/finishBundle as well.

Explode WindowedValues before processing.

sideOutputWithTimestamp to address start/finishBundle.

SideInput with windows.

Unused for now, remove.

Take sideInput window startegy into account, for combine as well.

reduce code duplication.

Spark combine support.

reuse code where possible.

Expose sideInputs and insertDefault in Combine.Globally for direct translation.

Direct translation of Combine.Globally into Spark's aggregate function.

Make default run with 4 cores by default - makes tests run with multiple threads, but not too many.

SideInputReader for the Spark runner.

A common abstraction for Keyed and Global implementation.

Implement Combine.Globally via Spark's aggregate.

runnable-on-service profile doesn't need pluginManagement.

Removing test as it does not follow a deterministic combine implementation.

Context reuse is mostly for testing. To avoid a test failure that will stop the context and fail all
following tests we need to recreate the context if it's stopped as well.

### CODE REVIEW

**Comment by amitsela** [comment]
**Note:**
This PR originally came to address "enabling" ROS tests after they were disabled.
It turned into re-implementing the runner to support the Beam model in full.

_Most_ of the work was on combine (key/globally) implementation to include `CombineWithContext`  support, but not only.

Another important note is testing: except for `testSessionsCombineWithContext` which is being discussed as [BEAM-696](https://issues.apache.org/jira/browse/BEAM-696) all ROS tests pass (batch).
With that in mind I would like to remove Spark's own custom tests that are covered by ROS. One thing comes to mind is the fact that `PAssert` now tests from within a DoFn and Spark's lazy nature will simply not run the DoFn if for some reason it won't ingest any input data (or drop something before the assert phase) and thus won't fail.

This is PR is a bit "bulky", but I hope that'll put the runner back on track, so excuse me for that.

**Comment by amitsela** [comment]
R: @kennknowles, @dhalperi 
Also @jbonofre PTAL

File: runners/spark/pom.xml [hunk 1]
```diff
@@ -59,40 +59,40 @@
...
-                      <beam.spark.test.reuseSparkContext>true</beam.spark.test.reuseSparkContext>
-                      <spark.ui.enabled>false</spark.ui.enabled>
-                    </systemPropertyVariables>
-                  </configuration>
-                </execution>
-              </executions>
-            </plugin>
-          </plugins>
-        </pluginManagement>
+        <plugins>
```
Comment (kennknowles) [inline runners/spark/pom.xml]: This seems unusual. Why is it needed?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java [hunk 2]
```diff
@@ -125,8 +164,50 @@ public InputT element() {
...
+      @Override
+      public BoundedWindow window() {
+        throw new UnsupportedOperationException("Window not available for "
+            + "start/finishBundle output.");
+      }
+    };
+    try {
+      @SuppressWarnings("unchecked")
+      Collection<? extends BoundedWindow> windows = windowFn.assignWindows(assignContext);
+      Instant outputTimestamp = timestamp != null ? timestamp : BoundedWindow.TIMESTAMP_MIN_VALUE;
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: I don't believe `timestamp` can ever be null, given that this is called from `outputWithTimestamp` and variants.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java [hunk 3]
```diff
@@ -18,76 +18,115 @@
...
+  Iterable<ValueT> callWithCtxt(Iterator<WindowedValue<InputT>> iter) throws Exception{
+    this.setup();
+    // skip if bundle is empty.
+    if (!iter.hasNext()) {
+      return Lists.newArrayList();
+    }
+    try {
+      fn.setup();
+      fn.startBundle(this);
+      return this.getOutputIterable(iter, fn);
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: Does this factoring mean that there might be a single `startBundle` call followed by multiple iterations, thus multiple `finishBundle` calls?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java [hunk 4]
```diff
@@ -125,8 +164,50 @@ public InputT element() {
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: Side note: line 152 (not part of this diff) has a spelling error of `AggregatprInputT`

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java [hunk 5]
```diff
@@ -125,8 +164,50 @@ public InputT element() {
...
+      output(WindowedValue.of(output, timestamp, windowedValue.getWindows(),
+          windowedValue.getPane()));
+    }
+  }
+
+  static <T> WindowedValue<T> noElementWindowedValue(final T output,
+                                                     final Instant timestamp,
+                                                     WindowingStrategy<?, ?> ws) {
+    @SuppressWarnings("unchecked")
+    WindowFn<Object, ?> windowFn = (WindowFn<Object, ?>) ws.getWindowFn();
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: From my reading, the only piece of `WindowingStrategy` used is the `WindowFn`. On the one hand, passing the whole `WindowingStrategy` makes it easy to drop new fields onto the parameter with backwards compatibility, but on the other hand, it means you have greater potential dependencies on the other fields.

File: runners/spark/pom.xml (hunk 1 above)
Comment (amitsela) [inline runners/spark/pom.xml]: Well, why use `pluginManagement` to begin with ? it doesn't "manage" any sub-modules, and I haven't seen this in other runners' pom.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java (hunk 3 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: Yes for a single `startBundle` followed by multiple iterations (processing elements in a partition of the micro-batch), but only one `finishBundle`, corresponding to the end-of-processing of this partition.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java (hunk 4 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: Thanks!

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java (hunk 5 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java]: You're right! it's leftover from a mid-work state before I used a dedicated `SideInputReader`, I needed it for the sideInputs, don't need it anymore. I'll pass the `WindowFn` only.

**Comment by amitsela** [comment]
I've addressed the comments so far, and pushed the applied changes. Keep'e coming 😉

**Comment by kennknowles** [comment]
Sorry for the delay!

The failures in Jenkins look real, but not related to the Spark runner nor to your changes outside of the runner. Perhaps they were known issues that are now fixed? I'd be curious if a rebase could clear up the Jenkins result.

I haven't had time to dig into the Spark details, but will do so tomorrow.

**Comment by amitsela** [comment]
I'll check why the IT test is failing, a for the unit test - still discussing this at [BEAM-696](https://issues.apache.org/jira/browse/BEAM-696)

**Comment by amitsela** [comment]
I've rebased on top of master, fixing a new JavaDoc issue, but nothing essential.

**Comment by amitsela** [comment]
It seems that now the `DataflowRunner` has fails on IT test, and the `DirectRunner` on a unit test.
The `SparkRunner` is not running ROS again, or it would have failed on BEAM-696.
Not sure what's going on here, I'm making a few more changes and we'll see how that goes,

**Comment by amitsela** [comment]
Still issue with the `DirectRunner`, though I clearly didn't change it or the test it fails `org.apache.beam.sdk.testing.PAssertTest.testContainsInAnyOrderFalse` and still ROS not running for the `SparkRunner`.
Is ROS supposed to run on PreCommit ?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java [hunk 6]
```diff
@@ -386,7 +386,7 @@ private static Throwable runExpectingAssertionFailure(Pipeline pipeline) {
     // is first caught by JUnit and causes a test failure.
     try {
       pipeline.run();
-    } catch (AssertionError exc) {
+    } catch (Exception exc) {
```
Comment (amitsela) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java]: The Spark runner will propagate the original error wrapped in a `RuntimeException`, see [here](https://github.com/apache/incubator-beam/blob/master/runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java#L172).

I could change this to catching a `Throwable`, which is also the return value of whatever is caught anyway. Also note that the message assertion passes so it does surface the original error.

WDYT ?

Comment (amitsela) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java]: I also don't think it'll break, if it catches an exception/throwable instead since it still asserts the expected message and it'll have to be a very weird coincidence to match a completely different error.

Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java]: Longer-term, I don't think it has to be the exact inner error, but I do think we need to standardize on the outermost error - users need to be able to switch runners via configuration and have their code (including try/catch of pipeline failures) keep working the same way.

Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java]: For now, catching a `Throwable` and checking the error message is OK. But I wonder if it might be easy to polish this. In fact, it is [`Pipeline.run()` that unwraps `UserCodeException`](https://github.com/apache/incubator-beam/blob/f62d04e22679ee2ac19e3ae37dec487d953d51c1/sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java#L183) and `PAssert` failures should be an `AssertionError` wrapped in a `UserCodeException` which the runner should let through.

I'm a bit confused about the full state of things here, but it might actually be worth a little thread on the dev list.

Comment (amitsela) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java]: I see your point, and I will take another look to see if I can change the propagated Exception to `UserCodeException` and check if that solves the issue. A rebase is in place anyway..

**Comment by amitsela** [comment]
I've rebased on master.
In addition, it appears that the runner's own `SparkProcessException` was designated to do what `UserCodeException` is doing (from spark-dataflow days..).
I replaced it to use `UserCodeException` and reverted back to `AssertionError` - this should make more sense now.

**Comment by amitsela** [comment]
The IT test failure looks like the known inconsistency issue mentioned in [BEAM-747](https://issues.apache.org/jira/browse/BEAM-747), and the unit test failure was due to a `BindException` on the embedded Kafka port.
Let's keep going, I'm sure I'll have more commits before this is LGTMed so the tests will run again.

One thing I don't understand is ROS tests - they're clearly not running or they would fail on `testSessionsCombineWithContext` specifically (known issue) - @kennknowles any chance you know why ?
This is a serious issue, last time those tests were off we missed-out on a lot of progress and it took me a while to catch-up. I really want to remove all of the runner's unit tests (batch, and in the future streaming) but I have to know that the ROS tests run on PreCommit.

**Comment by amitsela** [comment]
I've been working on utilizing `SimpleDoFnRunner` for the SparkRunner and it seems to be pretty straight-forward (except for some minor tweaks required due to the lazy/microbatch nature).
I think though that this PR is too extensive as it is so I will open a separate ticket let's concentrate on the combiners for now.
Thanks!

**Comment by kennknowles** [comment]
As for ROS tests, they run only in post-commit, both just for latency of results and because, taken all together, they would time out Jenkins and Travis.

**Comment by kennknowles** [comment]
With some Jenkins hackery to parallelize sub-builds, it seems feasible to:

1. Run unit tests
2. (sub tasks) Run ROS tests for each runner in parallel, skipping unit tests

I've set up Jenkins to do such things, so I know it can be done in general, but I'm not particularly close with the Apache Jenkins setup.

**Comment by amitsela** [comment]
As a general comment on ROS - if runners are to relay on them, how can we have them run only on post commit ? maybe only run the ROS on a specific runner if it's/SDK's code changed ?
So for examples commits to Flink runner won't run ROS on Spark runner (and vice versa).
WDYT ?

**Comment by kennknowles** [comment]
I agree that sounds pretty nice. I think the infrastructure issues involve some real work. This might be a good discussion for the dev list. I don't have a ton of insight into the ways we can configure things like that. The merge bot in development might help.

Once workflow I use a lot (when I can) is to separate (1) human feedback from (2) quick machine feedback from (3) slow machine feedback. Humans do what they are good at - style, architecture, sanity checking - and quick machine feedback handles minutia. The slower feedback can often wait until I've decided to submit (or I can run it myself if my change is sensitive to it). I think this workflow would manifest as shrinking Jenkins and Travis to much less, while having mergebot do much more.

**Comment by amitsela** [comment]
I'm afraid I'm no expert when it comes to testing/deployment infra. I just had a thought that seemed worth sharing.
I'll add this to my list of things that go in the Dev list 😉

**Comment by kennknowles** [comment]
There does seem to be a real failure at https://builds.apache.org/job/beam_PreCommit_MavenVerify/3995/org.apache.beam$beam-runners-spark/testReport/junit/org.apache.beam.runners.spark.translation.streaming/ResumeFromCheckpointStreamingTest/org_apache_beam_runners_spark_translation_streaming_ResumeFromCheckpointStreamingTest/ where Kafka is binding to a port that is in use (maybe conflicting with a concurrent build or test case?)

```
kafka.common.KafkaException: Socket server failed to bind to localhost:36520: Address already in use.
	at kafka.network.Acceptor.openServerSocket(SocketServer.scala:260)
	at kafka.network.Acceptor.<init>(SocketServer.scala:205)
	at kafka.network.SocketServer.startup(SocketServer.scala:86)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:99)
	at org.apache.beam.runners.spark.translation.streaming.utils.EmbeddedKafkaCluster.startBroker(EmbeddedKafkaCluster.java:125)
	at org.apache.beam.runners.spark.translation.streaming.utils.EmbeddedKafkaCluster.startup(EmbeddedKafkaCluster.java:115)
	at org.apache.beam.runners.spark.translation.streaming.ResumeFromCheckpointStreamingTest.init(ResumeFromCheckpointStreamingTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

**Comment by amitsela** [comment]
Yeah, my implementation of an embedded Kafka chooses a port randomly, this seems like purely bad luck.

**Comment by kennknowles** [comment]
Got it. Is it possible to bind the embedded Kafka to port `0` and then find out which it was assigned? I understand this might be a plumbing pain. I've run the following,

    mvn integration-test -Plocal-runnable-on-service-tests -pl runners/spark -am

which surfaces a failure in `CombineTest.testSessionsCombineWithContext()`. Is this related to [BEAM-696](https://issues.apache.org/jira/browse/BEAM-696)?

Travis is doing its usual ~50% green.

Overall, I prefer having a running test config with a failure to no actual tests running, and it is better for collaboration to get this in, so LGTM. I'll merge it and suggest to roll forwards with any further improvements and fixes.

**Comment by kennknowles** [comment]
Actually, could you rebase and maybe reword the first commit messages, then merge? The last commit seems like it could merge into the first nicely, and then the list in the first commit could be a little more readable as bullets or something.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java [hunk 7]
```diff
@@ -1417,6 +1417,20 @@ private Globally(String name, GlobalCombineFn<? super InputT, ?, OutputT> fn,
...
+     * Returns the side inputs used by this Combine operation.
+     */
+    public List<PCollectionView<?>> getSideInputs() {
+      return sideInputs;
+    }
+
+    /**
+     * Returns whether or not this transformation applies a default value.
+     */
+    public boolean isInsertDefault() {
```
Comment (amitsela) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java]: I know, I suck at naming names so I went with standard.

**Comment by amitsela** [comment]
I'll rebase, reword and then merge.
Yes the failing test is infamous [BEAM-696](https://issues.apache.org/jira/browse/BEAM-696) 😄

**Comment by amitsela** [comment]
As for the Kafka port, it happened a couple of times since we've setup Jenkins, and that's less than different flaking services in some of our tests.
Adding to that the fact that I'm going towards relaying on ROS (and IT tests) for testing the runner, I don't see a point in investing there, unless this becomes a recurring pain.

BTW, you think I should remove the runner's batch tests ? relay on ROS from now on ?

**Comment by amitsela** [comment]
I don't think the failure has anything to do with this PR but I'll kick Jenkins again just to be sure.

**Comment by amitsela** [comment]
I'll merge for now, filed [BEAM-781](https://issues.apache.org/jira/browse/BEAM-781) for removing the unit tests.
